### PR TITLE
[MXNET-317] Add Data Parallel

### PR DESCRIPTION
**Pull request by zhanghang1989**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Feature1, tests, (and when applicable, API doc)
- [x] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by zhanghang1989** [comment]
Hi @eric-haibin-lin @piiswrong , DataParallel and Barrier (ParallelState) are included as discussed.
Barrier is renamed => ParallelState, as suggested by @szha

File: python/mxnet/gluon/contrib/parallel.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        """Pull the output to each device
+        Input:
+            idx (int)
+
+        Output:
+            out (NDArray)
+        """
+        return self.out[idx]
+
+    def get(self):
```
Comment (eric-haibin-lin) [inline python/mxnet/gluon/contrib/parallel.py]: I think Sparse can reuse the pull() interface

Comment (zhanghang1989) [reply python/mxnet/gluon/contrib/parallel.py]: you mean removing get() function?

**Comment by zhanghang1989** [comment]
@pengzhao-intel This PR is mainly for Synchronized Cross GPU Batch Norm https://github.com/zhanghang1989/MXNet-Gluon-SyncBN

**Comment by zhanghang1989** [comment]
I have finished editing this repo. Could you start reviewing? @eric-haibin-lin @piiswrong Please see the deployed docs http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-10536/12/api/python/gluon/contrib.html?highlight=dataparallel#mxnet.gluon.contrib.parallel.DataParallelModel

File: python/mxnet/gluon/contrib/parallel.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    lock = threading.Lock()
+    results = {}
+
+    def _worker(i, module, input, kwargs, results, is_recording, is_training, lock):
+        try:
+            if is_recording:
+                with autograd.record(is_training):
+                    output = tuple_map(module(*input, **kwargs))
+                    for out in output:
+                        out.wait_to_read()
```
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply python/mxnet/gluon/contrib/parallel.py]: The training fails without waiting. I think it may because of `with autograd.record(is_training):`

File: docs/api/python/gluon/contrib.md [hunk 3]
```diff
@@ -80,6 +80,20 @@ In the rest of this document, we list routines provided by the `gluon.contrib` p
...
+#### Parallel
+
+```eval_rst
+.. currentmodule:: mxnet.gluon.parallel
+
+.. autosummary::
+    :nosignatures:
+    
+    DataParallelModel
+    DataParallelCriterion
```
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply docs/api/python/gluon/contrib.md]: This make it easier for the users, because the situation is complicated for network with multiple output cases. We just handle the situation internally.

File: python/mxnet/gluon/contrib/parallel.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+                self.all_tasks_done.notify_all()
+            while self.reduce_tasks:
+                self.all_tasks_done.wait()
+
+    def _clear(self):
+        self.list = []
+        self.push_tasks = self.counter
+        self.reduce_tasks = self.counter
+
+    def __len__(self):
```
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply python/mxnet/gluon/contrib/parallel.py]: We may want to get the number of devices outside. For example, we want to calculate the global mean by dividing the global sum by global number of elements (local number of elements * N devices)

File: python/mxnet/gluon/contrib/parallel.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+        return self.out[idx]
+
+    def _sync_op(self):
+        with self.mutex:
+            if self.reduce_tasks == 1:
+                assert(len(self.list) == self.counter)
+                self.out = self.op(*self.list)
+                if isinstance(self.out, (list, tuple)):
+                    for xi in self.out:
+                        xi.wait_to_read()
```
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply python/mxnet/gluon/contrib/parallel.py]: If we don't wait, the training of BN fails.

File: python/mxnet/gluon/contrib/parallel.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+                    out.wait_to_read()
+            with lock:
+                results[i] = output
+        except Exception as e:
+            with lock:
+                results[i] = e
+
+    is_training = autograd.is_training()
+    is_recording = autograd.is_recording()
+    threads = [threading.Thread(target=_worker,
```
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply python/mxnet/gluon/contrib/parallel.py]: I mainly followed the PyTorch implementation. Could you show an example of improvement?

File: python/mxnet/gluon/contrib/parallel.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+        The cross device operation is applying (e.g. AllReduce).
+    """
+    def __init__(self, counter, operation):
+        self.mutex = threading.Lock()
+        self.all_tasks_done = threading.Condition(self.mutex)
+        self.counter = counter
+        self.op = operation
+        self._clear()
+
+    def push(self, x):
```
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply python/mxnet/gluon/contrib/parallel.py]: There was a case in 'Sparse' application, which only pull onetime from the master thread. We can change it to single function.

File: python/mxnet/gluon/contrib/parallel.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+            if self.push_tasks == 0:
+                self._clear()
+            self.list.append(x)
+            idx = len(self.list) - 1
+            self.push_tasks -= 1
+
+        with self.all_tasks_done:
+            if self.push_tasks == 0:
+                self.all_tasks_done.notify_all()
+            while self.push_tasks:
```
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply python/mxnet/gluon/contrib/parallel.py]: avoid execute the `self.op` before all the datas are ready, see the `_sync_op()` function

File: python/mxnet/gluon/contrib/parallel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (piiswrong) [reply python/mxnet/gluon/contrib/parallel.py]: Why? What's the error? That shouldn't happen. Are you using most recent code? There was a fix for multithreading

File: python/mxnet/gluon/contrib/parallel.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (piiswrong) [reply python/mxnet/gluon/contrib/parallel.py]: Use a thread pool at constructor and reuse it for each iteration?

If pytorch do it like this maybe its fine

File: python/mxnet/gluon/contrib/parallel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply python/mxnet/gluon/contrib/parallel.py]: I haven't tested recently, I will take a look.

File: docs/api/python/gluon/contrib.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (piiswrong) [reply docs/api/python/gluon/contrib.md]: I think this is too many API's and prone to misuse.

Also we use the term Loss instead of Criterion

File: python/mxnet/gluon/contrib/parallel.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+        self.module = module
+        self.sync = sync
+
+    def __call__(self, *inputs, **kwargs):
+        if not self.ctx_list:
+            return self.module(*inputs, **kwargs)
+        inputs, kwargs = _split_load_kwargs(inputs, kwargs, self.ctx_list)
+        assert(len(inputs) == len(self.ctx_list))
+        if len(self.ctx_list) == 1:
+            return tuple([tuple_map(self.module(*inputs[0], **kwargs[0]))])
```
Comment (piiswrong) [inline python/mxnet/gluon/contrib/parallel.py]: (x,) makes a tuple

File: docs/api/python/gluon/contrib.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply docs/api/python/gluon/contrib.md]: Will change the name :)

File: tests/python/unittest/test_contrib_parallel.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        # The Flatten layer collapses all axis, except the first one, into one axis.
+        net.add(nn.Flatten())
+        net.add(nn.Dense(512,in_units=800))
+        net.add(nn.Activation('relu'))
+        net.add(nn.Dense(10, in_units=512))
+
+    net.collect_params().initialize()
+    criterion = gluon.loss.SoftmaxCELoss(axis=1)
+
+    def test_net_sync(net, criterion, sync, nDevices):
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_contrib_parallel.py]: nDevices -> num_devices

File: tests/python/unittest/test_contrib_parallel.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+    class MyLayer(Block):
+        def __init__(self, nGPU):
+            super(MyLayer, self).__init__()
+            self.barrier = Barrier(nGPU, my_callable)
+
+        def forward(self, x):
+            idx = self.barrier.push(x)
+            y = self.barrier.pull(idx)
+            assert_allclose(y.asnumpy(), x.asnumpy(), rtol=1e-2, atol=1e-4)
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_contrib_parallel.py]: should x and y be exactly the same? The tolerance seems large

File: python/mxnet/gluon/contrib/parallel.py [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+        return len(self.list)
+
+    def __repr__(self):
+        return 'Barrier'
+
+
+class DataParallelModel(object):
+    """Data Parallelism
+
+    Hide the difference of single/multiple GPUs to the user.
```
Comment (eric-haibin-lin) [inline python/mxnet/gluon/contrib/parallel.py]: Hide .. from ..

File: python/mxnet/gluon/contrib/parallel.py [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+    the batch dimension.
+    In the forward pass, the module is replicated on each device,
+    and each replica handles a portion of the input. During the backwards pass,
+    gradients from each replica are summed into the original module.
+    Note that the outputs are not gathered, please use compatible
+    :class:`mxnet.gluon.contrib.DataParallelModel`
+
+    Parameters
+    ----------
+    module : object
```
Comment (eric-haibin-lin) [inline python/mxnet/gluon/contrib/parallel.py]: Is it better to call it block instead of module?

File: tests/python/unittest/test_contrib_parallel.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    iters = 100
+    # train mode
+    for i in range(iters):
+        x = mx.random.uniform(shape=(8, 1, 28, 28))
+        with autograd.record():
+            y = net(x)
+    # evaluation mode
+    for i in range(iters):
+        x = mx.random.uniform(shape=(8, 1, 28, 28))
+        y = net(x)
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_contrib_parallel.py]: what specifically are you checking here?

File: python/mxnet/gluon/contrib/parallel.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply python/mxnet/gluon/contrib/parallel.py]: Is the async execution engine not able to handle this?

File: python/mxnet/gluon/contrib/parallel.py (hunk 13 above)
Comment (zhanghang1989) [reply python/mxnet/gluon/contrib/parallel.py]: Block has a general meaning in Gluon. We can probably call it model, if you think that is better.

File: tests/python/unittest/test_contrib_parallel.py (hunk 14 above)
Comment (zhanghang1989) [reply tests/python/unittest/test_contrib_parallel.py]: mainly check the behavior of Barrier:
``assert_allclose(y.asnumpy(), x.asnumpy(), rtol=1e-2, atol=1e-4)``

File: python/mxnet/gluon/contrib/parallel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhanghang1989) [reply python/mxnet/gluon/contrib/parallel.py]: Actually, people can use ``sync=False`` mode at most of the cases.

**Comment by zhanghang1989** [comment]
Closing it due to no longer needed in SyncBN https://github.com/apache/incubator-mxnet/pull/11502
